### DISCUSSION TITLE
Pros/Cons of client-side vs server-side database requests

### DISCUSSION THREAD

**Discussion by seb-lewis** [sveltejs/kit#7438] [Q&A]
Just wondering, is there any particular benefit/security concerns that would make it better to have server routes fetch data from the database before serving it to the client? Or is it fine to use something like Supabase on the client to fetch and process database data?

Thanks in advance for the clarification!

**Comment by MentalGear** [comment]
You'd need to check with the db provider and their SDK what mode of operation (and exposition) is safe, along with your database auth setup.

For supabase, if you want to have the client communicate directly with your supabase instance, you must set a RLS policy as your anon key will be exposed client-side.

Please check the supabase SDK/js/auth-helpers docs for more details.

**Reply by RykGitProjects** [reply]
Is it faster to user +page.js instead of +page.server.js ? I also use Supabase

**Reply by david-plugge** [reply]
> Is it faster to user +page.js instead of +page.server.js ? I also use Supabase

If you send multiple queries to supabase you are probably better of using the server as it will have better internet speed than your clients.
That way the client only makes a single request per page and your server calls your api multiple times. In addition you can also cache the data fetched from supabase when using `+page.server.js`.

If you want to build an spa or even a pwa i´d suggest `+page.js` as you don´t even need the sveltekit server once the pwa is installed

**Reply by RykGitProjects** [reply]
Ok got it

And how about getting data in onMount (example: https://languageimperfect.com/2021/02/17/data-fetching-in-svelte.html)

This way user will not wait for document to load, which takes a couple of hunderd miliseconds

**Reply by david-plugge** [reply]
That should be pretty straight forward as you only fetch on the client in that case. You can simply use the global supabaseClient.

I´d suggest this though:

```ts
// src/routes/+page.js
import { writable } from 'svelte/store';
import { browser} from '$app/environment';

export async function load(event) {
	const posts = writable([]);

	// only call the api on the client
	// but dont await the promise so the page is instantly shown
	// and the store is lazily updated with the latest data from supabase
	if (browser) {
		const { supabaseClient } = await getSupabase(event);

		supabaseClient
			.from('posts')
			.select()
			.then(({ data }) => {
				posts.set(data);
			});
	}

	return {
		posts
	}
}
```
```html
<script>
	export const data;

	$: posts = data.posts
</script>

{#each $posts as post}
	{post.title}
{/each}
```

This way we don´t call supabase on the server, instantly show the requested page (lazily load the page data) but keep the benefit of prefetching when hovering a link https://kit.svelte.dev/docs/link-options#data-sveltekit-prefetch

**Reply by RykGitProjects** [reply]
Great thank you!

From your perspective which way of getting data from Supabase is better: +page.server.js or on client as in your code?

As far as I understand the biggest advantage of getting it on client is that there is much smaller document to wait for, so page can have some spinner when waiting for data to load, instead of "blank page"

**Reply by david-plugge** [reply]
That depends on the type of app you are building.
If you care about seo, you want to avoid getting data in your page components (basicly you want full ssr). If you are don't care about a loading spinner on the first page load then the example I gave should be a pretty good start. The nice thing with svelte kit is that you can go with ssr for some part of your app while some other pages are csr only.

**Reply by david-plugge** [reply]
On the road to 1.0 we are probably getting more examples for the different use cases.
I was thinking about creating a more complex application with supabase and svelte kit as an entry point

**Reply by seb-lewis** [reply]
> On the road to 1.0 we are probably getting more examples for the different use cases.
> 
> I was thinking about creating a more complex application with supabase and svelte kit as an entry point

That would be awesome. I'm also curious about best practices for things like pagination and keeping client data up to date after there's been an update from the frontend and/or backend.

**Reply by RykGitProjects** [reply]
> That depends on the type of app you are building. If you care about seo, you want to avoid getting data in your page components (basicly you want full ssr). If you are don't care about a loading spinner on the first page load then the example I gave should be a pretty good start. The nice thing with svelte kit is that you can go with ssr for some part of your app while some other pages are csr only.

SEO is really important for me. I assumed, that for google there is no difference. If Search Console reads page properly (checking URL inspection) with all data, then there is no difference. If there is I definitely need to keep +page.server.js

**Reply by david-plugge** [reply]
> SEO is really important for me. I assumed, that for google there is no difference. If Search Console reads page properly (checking URL inspection) with all data, then there is no difference. If there is I definitely need to keep +page.server.js

You don´t necessarily need the `+page.server.js` as `+page.js` load functions are executed on the server and the client ([docs](https://kit.svelte.dev/docs/load#shared-vs-server)).
I´m not an seo expert at all so i may be mistaken, but i think you are good when all publicly available data is server rendered. User specific data can´t be indexed anyway so you should be good loading that on the client only. There are some docs about [seo](https://kit.svelte.dev/docs/seo) in sveltekit you may want to take a look at (they actually seem to be outdated, you can´t disable ssr within handle anymore i think).

You can also modify the code i posted to enable ssr and fall back to lazily loading data on the client
```js

export async function load(event) {
	const posts = writable([]);

	const { supabaseClient } = await getSupabase(event);

	const postPromise = supabaseClient
		.from('posts')
		.select()
		.then(({ data }) => {
			posts.set(data);
		});

	// await the promise if we are on the server to enable ssr
	if (!browser) {
		await postPromise;
	}

	return {
		posts
	}
}
```

**Reply by david-plugge** [reply]
> That would be awesome. I'm also curious about best practices for things like pagination and keeping client data up to date after there's been an update from the frontend and/or backend.

Pagination with supabase or in general with sveltekit?

**Reply by RykGitProjects** [reply]
Tried loading data lazily on front (so front is not waiting for data to be gathered). Seems like supabase is unable to understand this type of connection. My code gets me an TypeError, tried following 1:1:

```
// src/routes/postlists/+page.js
import { writable } from 'svelte/store';
import { browser} from '$app/environment';
import { getSupabase } from '../../supabaseImport'

export async function load(event) {
	const posts = writable([]);

	// only call the api on the client
	// but dont await the promise so the page is instantly shown
	// and the store is lazily updated with the latest data from supabase
	if (browser) {
		const { supabaseClient } = await getSupabase(event);

		supabaseClient
			.from('my_table')
			.select('column_with_posts')
			.then(({ data }) => {
				posts.set(data);
			});
	}

	return {
		posts
	}
}
```

```
// src/routes/postlists/+page.svelte
<script>
	export let data;

	$: posts = data.posts
        console.log(posts)
</script>

```
```

//supabaseImport.js
import { createClient } from '@supabase/supabase-js'

const supabaseUrl = import.meta.env.VITE_SUPABASE_URL
const supabaseAnonKey = import.meta.env.VITE_SUPABASE_ANON_KEY

export const supabase = createClient(supabaseUrl, supabaseAnonKey)
```

**Reply by david-plugge** [reply]
In the example i gave i assumed you were using the [supabase auth helpers](https://github.com/supabase/auth-helpers/blob/main/packages/sveltekit/README.md)

**Reply by RykGitProjects** [reply]
Thx, will try it !

**Comment by seb-lewis** [comment]
I guess in general works also, but supabase specifically would be great.

**Reply by david-plugge** [reply]
You probably want to edit that message and remove the unsubscribe link
